### PR TITLE
PENG-2369 implement the license manager agent snap

### PR DESCRIPTION
**Pull request by matheushent**
#### What
This PR implement a snap for installing and running the license manager agent. As well as, it modifies the workflows so quality assurance is run against the snap sub project. Also, a new job is added in the *Publish on Tag* workflow for publishing the snap to the *latest/candidate* and *latest/edge* channels.

#### Why
This PR facilitates the installation and deployment of the agent

`Task`: https://sharing.clickup.com/t/h/c/18022949/PENG-2369/RLHZFG26P6LMK8P

---

#### Peer Review
Please follow the upstream omnivector documentation concerning
[peer-review guidelines](https://github.com/omnivector-solutions/Documentation/blob/main/Contributing/pr_review_standards.md#peer-review).

### CODE REVIEW

File: lm-agent-snap/Makefile [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+.PHONY: type
+type: ## Run the type checker
+	tox -e type
+
+.PHONY: qa
+qa: lint type ## Run all quality assurance checks
+
+.PHONY: format
+type: ## Run the formatter
```
Comment (dusktreader) [inline lm-agent-snap/Makefile]:
```suggestion
format: ## Run the formatter
```

File: lm-agent-snap/README.md [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+# Installation instructions
+
+For installing from the remote store, run:
+```bash
+sudo snap install license-manager-agent
+```
+
+## Basic Usage
+
+This snap requires a few configuration values to be set before it can be used. These values are:
```
Comment (dusktreader) [inline lm-agent-snap/README.md]: I think we should change the wording here since most of these settings are not required to be changed. It might be good to break up the settings into two groups. The first could be the required settings (`oidc-client-id` and `oidc-client-secret`) and the second list could be the optional settings. What do you think?

File: lm-agent-snap/README.md [hunk 3]
```diff
@@ -0,0 +1,76 @@
+# lm-agent-snap
+The License Managar Agent bundled into a Snap.
+
+# Installation instructions
+
+For installing from the remote store, run:
```
Comment (dusktreader) [inline lm-agent-snap/README.md]: What do you think of having this say "Snap store" instead and make it a link?

File: lm-agent-snap/README.md [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+
+- scontrol-path: The absolute path to the *scontrol* command on the host system. This is optional and defaults to /usr/bin/scontrol.
+
+- default-slurm-work-dir: The default working directory that the agent will use when submitting jobs to the SLURM cluster. This is optional and defaults to /tmp.
+
+- slurm-user-mapper: The user mapper that the agent will use to map the system user name to the SLURM user name. This is optional and defaults to none.
+
+- single-user-submitter: The system user name that the agent will use to submit jobs to the SLURM cluster on behalf of in case the *single-user-submitter* is not set. This is optional and defaults to *ubuntu*.
+
+- write-submission-files: A boolean value (true, false) that indicates whether the agent should write submission files to disk. This is optional and defaults to false.
```
Comment (dusktreader) [inline lm-agent-snap/README.md]: This needs to be re-written for the license manager config settings.

File: lm-agent-snap/README.md [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+Any configuration can be set using the *snap* command line, e.g.:
+```bash
+sudo snap set license-manager-agent oidc-client-id=foo
+sudo snap set license-manager-agent oidc-client-secret=boo
+```
+
+# Development
+
+For development purposes, you can build the `license-manager-agent` part prior of packing the snap. To do that, run:
```
Comment (dusktreader) [inline lm-agent-snap/README.md]:
```suggestion
For development purposes, you can build the `license-manager-agent` part prior to packing the snap. To do that, run:
```

File: lm-agent-snap/hooks/bin/configure [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+    snapctl_out: Union[str, None]
+    snapctl_out = run_bash(f"snapctl get {snap_config_value}")
+
+    if snapctl_out == "":
+        snapctl_out = None
+
+    return snapctl_out
+
+
+def run_bash(bash_string: str) -> str:
```
Comment (dusktreader) [inline lm-agent-snap/hooks/bin/configure]: It would be good to move this to the top of the file since it is used by functions that are defined before it.

File: lm-agent-snap/hooks/bin/configure [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+    except Exception:
+        sys.exit(1)
+
+
+def snapctl_get(snap_config_value: str) -> Union[str, None]:
+    """Get snap config from snapctl.
+
+    Return python None if snapctl returns the empty string.
+    """
+    snapctl_out: Union[str, None]
```
Comment (dusktreader) [inline lm-agent-snap/hooks/bin/configure:61]: Why use the old style type definintion here instead of `str | None`?

File: lm-agent-snap/snap/snapcraft.yaml [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+
+  - lsdyna-path: Absolute path to the binary for lstc_qrun (needed for LS-Dyna licenses). Defaults to /usr/local/bin/lsdyna.
+
+  - lmxendutil-path: Absolute path to the binary for lmxendutil (needed for LM-X licenses). Defaults to /usr/local/bin/lmxendutil.
+
+  - olixtool-path: Absolute path to the binary for olixtool (needed for OLicense licenses). Defaults to /usr/local/bin/olixtool.
+
+  - reservation-identifier: Reservation name used when doing the reconciliation. Defaults to *license-manager-reservation*.
+
+  - lm_user: System user name for creating reservations on behalf of. Defaults to *license-manager*.
```
Comment (dusktreader) [inline lm-agent-snap/snap/snapcraft.yaml]:
```suggestion
  - lm_user: System user name used when creating reservations. Defaults to *license-manager*.
```

**Review by dusktreader — CHANGES_REQUESTED** [review]

File: lm-agent-snap/README.md (hunk 2 above)
Comment (matheushent) [reply lm-agent-snap/README.md]: that's a good idea, thank you

File: lm-agent-snap/README.md (hunk 4 above)
Comment (matheushent) [reply lm-agent-snap/README.md]: damn, I forgot to copy from the snapcraft.yaml file

File: lm-agent-snap/hooks/bin/configure (hunk 7 above)
Comment (matheushent) [reply lm-agent-snap/hooks/bin/configure:61]: I don't know why but the type checker failed sometimes for the vantage agent. Here I just copied tbh.

Comment (matheushent) [reply lm-agent-snap/hooks/bin/configure:61]: I also remember having errors after the snap got installed due to python versions not accepting this. For example, installing the snap on ubuntu 20.04 would fail

File: lm-agent-snap/snap/snapcraft.yaml (hunk 8 above)
Comment (matheushent) [reply lm-agent-snap/snap/snapcraft.yaml]: modified in [13c89e3](https://github.com/omnivector-solutions/license-manager/pull/389/commits/13c89e37e4fca4d4f8cd3de0090c4f227cc7a738)

File: lm-agent-snap/hooks/bin/configure (hunk 6 above)
Comment (matheushent) [reply lm-agent-snap/hooks/bin/configure]: modified in [13c89e3](https://github.com/omnivector-solutions/license-manager/pull/389/commits/13c89e37e4fca4d4f8cd3de0090c4f227cc7a738)

File: lm-agent-snap/README.md (hunk 5 above)
Comment (matheushent) [reply lm-agent-snap/README.md]: done in [13c89e3](https://github.com/omnivector-solutions/license-manager/pull/389/commits/13c89e37e4fca4d4f8cd3de0090c4f227cc7a738)

File: lm-agent-snap/README.md (hunk 4 above)
Comment (matheushent) [reply lm-agent-snap/README.md]: fixed in [13c89e3](https://github.com/omnivector-solutions/license-manager/pull/389/commits/13c89e37e4fca4d4f8cd3de0090c4f227cc7a738)

File: lm-agent-snap/README.md (hunk 3 above)
Comment (matheushent) [reply lm-agent-snap/README.md]: done in [13c89e3](https://github.com/omnivector-solutions/license-manager/pull/389/commits/13c89e37e4fca4d4f8cd3de0090c4f227cc7a738)

File: lm-agent-snap/README.md (hunk 2 above)
Comment (matheushent) [reply lm-agent-snap/README.md]: done in [13c89e3](https://github.com/omnivector-solutions/license-manager/pull/389/commits/13c89e37e4fca4d4f8cd3de0090c4f227cc7a738)

File: lm-agent-snap/Makefile (hunk 1 above)
Comment (matheushent) [reply lm-agent-snap/Makefile]: fixed in [13c89e3](https://github.com/omnivector-solutions/license-manager/pull/389/commits/13c89e37e4fca4d4f8cd3de0090c4f227cc7a738)

File: lm-agent-snap/hooks/bin/configure (hunk 7 above)
Comment (dusktreader) [reply lm-agent-snap/hooks/bin/configure:61]: I guess this must be due to the snap using the system python. For 20.04, it's Python 3.8. So, this would explain why there are problems. Thanks for the explanation!

**Review by dusktreader — APPROVED** [review]

File: lm-agent/lm_agent/config.py [hunk 9]
```diff
@@ -10,7 +10,7 @@
 from lm_agent.constants import LogLevelEnum
 
 
-DEFAULT_DOTENV_PATH = Path("/etc/default/license-manager-agent")
+DEFAULT_DOTENV_PATH = Path("/var/snap/license-manager-agent/common/.env")
```
Comment (julianaklulo) [inline lm-agent/lm_agent/config.py:13]: Is this change really necessary?
In Scania's environment the `.env` file is loaded from the default path, the charm is configured to populate the env vars in there.

**Review by julianaklulo — CHANGES_REQUESTED** [review]

File: lm-agent/lm_agent/config.py (hunk 9 above)
Comment (matheushent) [reply lm-agent/lm_agent/config.py:13]: this is needed according to our intention in making the agent installable by snap instead of configuring the virtual environment, installing the python package, etc...

Comment (dusktreader) [reply lm-agent/lm_agent/config.py:13]: Damn...Juliana is right, this will cause issues for our deployments of lm-agent on non-Ubuntu machines. We have some deployments on Rocky 9 and those environments do not have snap installed (and we will not be able to install it, either).

Instead of having one path we look in for the dotenv file, perhaps we need to have a set of places that we look in order until we find a dotenv file. What do you both think of this?

Comment (matheushent) [reply lm-agent/lm_agent/config.py:13]: we discussed in private that we would move with this implementation. Can we resolve @dusktreader @julianaklulo ?

Comment (dusktreader) [reply lm-agent/lm_agent/config.py:13]: As per our prior discussion, the charm-license-manager-agent is already configuring the systemd service to use an `EnvironmentFile`. We verified that agent when deployed with the charm is able to load its config successfully without needing to set the DEFAULT_DOTENV_PATH to `/etc/default/license-manager-agent`. Resolving this comment since we are moving forward as planned and tested.

**Review by julianaklulo — APPROVED** [review]
